lldb/source/Host/common/File.cpp

```diff
@@ -274,7 +276,23 @@ int NativeFile::GetDescriptor() const {
 }
 
 IOObject::WaitableHandle NativeFile::GetWaitableHandle() {
+#ifdef _WIN32
+  return (HANDLE)_get_osfhandle(GetDescriptor());
+#else
   return GetDescriptor();
+#endif
+}
+
+bool NativeFile::HasReadableData() {
+#ifdef _WIN32
+  DWORD available_bytes = 0;
+  return !PeekNamedPipe((HANDLE)_get_osfhandle(GetDescriptor()), NULL, 0, NULL,
```

Comment:
This will only work for "files" that are actually pipes, right?

If so, I don't think that's a good long-term direction. My medium-to-long term plan would be to make Pipe an IOObject, but that requires splitting it into two (one IOObject for each end). I don't think that would be a particularly hard refactor, but if you don't want to take that on, maybe we can find a way to pass a Pipe to MainLoop directly (and keep NativeFile out of it)? I'd be fine with an RegisterReadObject overload that takes a Pipe (and ignores the write end of the pipe -- we don't  select on writes anyway).

Reply:
I've split this up for checking between files and pipes. For files, we'll get notified anytime the file changes on disk, for pipes we'll use the zero byte read in a monitor thread to signal when new data is available. That strategy should work with both named pipes and anonymous pipes. I'm checking the file type we get out of the IOObject though, so this should work if you passed in `stdin` as well.